### ISSUE TITLE
Problems with aiorabbit and missed heartbeats

### ISSUE THREAD

**Issue by mgrandi** [gmr/aiorabbit#6]
python version: `Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)]`
aiorabbit version: `1.0.0a1`

 I have a small script that is essentially doing the following, with aiorabbit, the first iteration was using `aiorabbit.connect() `

```plaintext

async with aiorabbit.connect() as client:

  for i in range(x,y):
    await client.publish()

```

however, when i use `aiorabbit.connect() `, i keep getting errors where it ends up throwing an exception "No heartbeat in 150.047000 seconds"

I"m not sure what I"m doing wrong, I'm basically following the example on the website, and another script that is using the callback method in "basic_consume" and i don't get any heartbeat errors with that script. 

I looked in the source code and saw that it takes a heartbeat query parameter , so i even passed a `?heartbeat=60` in the rabbitmq URL, didn't seem to do anything

snippets from the log:

```plaintext

2021-05-16T21:48:09.325902-07:00 MainThread aiorabbit.channel0                       DEBUG   : Checking for heartbeats every 60.000000 seconds

2021-05-16T21:48:39.316379-07:00 MainThread aiorabbit.channel0                       DEBUG   : Set state to 23: Heartbeat Received while state is 22: OpenOk Received - {22: {}, 34: {}} [None]
2021-05-16T21:48:39.317385-07:00 MainThread aiorabbit.channel0                       DEBUG   : Transition to 23: Heartbeat Received from 22: OpenOk Received after 30.0000 seconds
2021-05-16T21:48:39.320401-07:00 MainThread aiorabbit.channel0                       DEBUG   : Set state to 24: Heartbeat Sent while state is 23: Heartbeat Received - {22: {}, 34: {}} [None]
2021-05-16T21:48:39.321392-07:00 MainThread aiorabbit.channel0                       DEBUG   : Transition to 24: Heartbeat Sent from 23: Heartbeat Received after 0.0000 seconds



2021-05-16T21:51:09.370997-07:00 MainThread aiorabbit.channel0                       CRITICAL: No heartbeat in 150.047000 seconds
2021-05-16T21:51:09.371967-07:00 MainThread aiorabbit.client                         INFO    : Remote server closed the connection (599) Too many missed heartbeats
2021-05-16T21:51:09.371967-07:00 MainThread aiorabbit.client                         DEBUG   : Set state to 259: Closed while state is 112: Received message acknowledgement - {49: {}, 34: {}, 112: {}, 132: {}, 147: {}} [ConnectionClosedException('Too many missed heartbeats')]


```

any idea?

**Comment by mgrandi** [comment]
So i have done some debugging, and it is the library itself that is killing the connection because of 'missed heartbeats', but i think the detection logic is flawed

```python
    def _heartbeat_check(self):
        threshold = self._loop.time() - (self._heartbeat_interval * 2)
        if self._last_heartbeat < threshold:
            msg = 'No heartbeat in {:2f} seconds'.format(
                self._loop.time() - self._last_heartbeat)
            self._logger.critical(msg)
            self._heartbeat_timer = None
            self._on_remote_close(599, 'Too many missed heartbeats')
        else:
            self._heartbeat_timer = self._loop.call_later(
                self._heartbeat_interval, self._heartbeat_check)
```

However, according to https://www.rabbitmq.com/heartbeats.html, 

"Any traffic (e.g. protocol operations, published messages, acknowledgements) counts for a valid heartbeat. Clients may choose to send heartbeat frames regardless of whether there was any other traffic on the connection but some only do it when necessary."

It seems that aiorabbit is not sending any heartbeat frames itself, but only reads to when they get sent. This combined with the fact that rabbitMQ will not send any heartbeat frames to the client if there is traffic on that channel, i tested it out by having my exact same code but doing an `await asyncio.sleep(1)` instead of my `client.publish()` calls and i could see that rabbitMQ sends heartbeat frames to check to make sure that the client is still alive.

BUT, since the internal logic that aiorabbit uses requires the server to send heartbeat frames to reset the `self._last_heartbeat ` value, and since that wont get set if there is active traffic on the channel (and the rabbit server doesn't send heartbeat frames), aiorabbit thinks there hasn't been any heartbeat traffic and will kill the connection

so i guess the strategy is to make it so that aiorabbit considers all traffic and sets the `self._last_heartbeat ` value accordingly, or makes it so it sends heartbeat frames itself and doesn't rely on rabbitmq to send it

**Comment by gmr** [comment]
I've updated the logic to set the last heartbeat received value any time a frame is received. While it will create some additional overhead, I think it will prove useful in this case.

I'm not convinced this is at the root of the issue you are seeing. I expect this might have more to do with the bug that I fixed in #8 where the initial last heartbeat value was 0 and used to check if a heartbeat was missed.

Sorry for the delay in addressing this.
